README.md

```diff
@@ -93,20 +93,38 @@ repositories are defined in your build.gradle.
 
 ## Dependencies on other Jenkins Plugins
 
-If your plugin depends on other Jenkins plugins you can specify the dependencies in the following way:
+If your plugin depends on other Jenkins plugins, you can use the same _configurations_ as in Gradle's `java-libary` plugin.
+See [the documentation](https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_separation) for details on the difference of `api` and `implementation` dependencies.
+For _optional dependencies_, you can use Gradle's [feature variants](https://docs.gradle.org/current/userguide/feature_variants.html).
 
+You can define both dependencies to Jenkins plugins and plain Java libraries.
+The JPI plugin will figure out what you are depending on and process it accordingly (Java libraries will be packaged in the your Jenkins plugin's hpi/jpi file).
+
+The additional `jenkinsServer` configuration can be used to install extra plugins for the `server` task (see below).
+
+Examples:
+
+    java {
+        // define features for 'optional dependencies'
+        registerFeature('ant') {
+            usingSourceSet(sourceSets.main)
+        }
+    }
+    
     dependencies {
-        jenkinsPlugins 'org.jenkinsci.plugins:git:1.1.15'
-        optionalJenkinsPlugins 'org.jenkins-ci.plugins:ant:1.2'
-        jenkinsTest 'org.jenkins-ci.main:maven-plugin:1.480'
+        implementation 'org.jenkinsci.plugins:git:1.1.15'
+        api 'org.jenkins-ci.plugins:credentials:1.9.4'
+        
+        // dependency of the (optional) ant feature
+        antImplementation 'org.jenkins-ci.plugins:ant:1.2'
```

Comment:
I'm not sure feature variants quite capture the same thing as the `optionalJenkinsPlugins` configuration was getting at when declaring [optional plugins in Jenkins](https://jenkins.io/doc/developer/plugin-development/optional-dependencies/)?

> It is possible to declare dependencies to other plugins as optional. In this way, even if some of the optional dependencies are unavailable, the plugin will continue to be loaded and executed.

> You can mark your extension as an optional extension by changing the `@Extension` annotation to `@Extension(optional = true)` and Jenkins doesn’t log any class loading errors when reading it.



But in [Modeling feature variants and optional dependencies](https://docs.gradle.org/current/userguide/feature_variants.html), I see:

> Feature variants let consumers choose what features of a library they need: the dependency management engine will select the right artifacts and dependencies.

An example from the Jenkins plugins I've written is declaring an optional dependency on the [Build Trigger Badge](https://plugins.jenkins.io/buildtriggerbadge) plugin. The classes from Build Trigger Badge are available for compilation, but stack traces aren't logged if that plugin is missing. Since Jenkins is managing the plugins at runtime, it could become available (or unavailable) at any time by user action.

Reply:
In general, I used feature variants as it is a core concept of Gradle's Java support now. So I think in the long term it is easier to understand how to use the plugin if it follows standard Gradle features. So users who already know the concept immediately know how to use it here. It is **the way** to do "optional dependencies" in Gradle - see also: https://blog.gradle.org/optional-dependencies

Another option could be that the JPI plugin always introduces one _feature variant_ by default which is then used for optional dependencies (similar to the `optionalJenkinsPlugins` configuration, but with a feature variant backing it). Which is what you also propose - _a single registered feature called 'optional'_

Going to your example, A could depend on B and also on it's optional feature (lets call it F).

```
dependencies {
  implementation("B:1")
  implementation("B:1") {
    capabilities { requireCapability("B:F")
  }
}
```

Then Gradle would always have C as transitive dependency (as it is a dependency of B:F). But Jenkins itself won't pick that up. Right. But I think I could fix that in `DependencyAnalysis` - so if the dependency is to a feature variant, we would add all the dependencies of that variant as well to the Manifest.

So the manifest of A in the above case would be:
```
B:1,C:1
```
(At the moment I think it is `B:1`,`B:1` which is actually not right. I just missed this.)

And that would be the advantage of using feature variants then.

Should we do it that way? Wdyt @wolfs?

Reply:
I think it is better if the plugins manage to declare their feature variants properly. We could leave in the `optionalJenkinsPlugins` so the migration to the new plugin version is easier, though I would deprecate it and remove it in the long term.

The reasons for not adding a default feature variant is the same as why we don't add a default feature variant in the JVM plugins of Gradle: you shouldn't use it.

Reply:
I fixed the implementation. This really was an oversight. Thanks for bringing this up @sghill.

Now the dependency information that is sourced from an optional feature that was selected is automatically added to the JPI manifest so Jenkins will pick it up at runtime. As in this test:

https://github.com/jenkinsci/gradle-jpi-plugin/pull/132/files#diff-717dc1bed1b536b60ee52ba8cb3a59f7R177-R224

Reply:
Thanks for the update to the manifest and walking through the options. I feel more confident about the changes in place and the approach taken now. 

> I think in the long term it is easier to understand how to use the plugin if it follows standard Gradle features. 

I agree completely @jjohannes :+1:

> We could leave in the optionalJenkinsPlugins so the migration to the new plugin version is easier, though I would deprecate it and remove it in the long term.

I added a comment here about revisiting this after attempting to convert some consumers: https://github.com/jenkinsci/gradle-jpi-plugin/pull/132#discussion_r372559202

I think this may be a minor oversight of declaring the dependency twice? I imagine it's not required to declare the dependency with and without the `requireCapability`?

https://github.com/jenkinsci/gradle-jpi-plugin/blob/4fb6d2f5d41ac828ad50a4b7dc6988e0e998bfb0/src/test/groovy/org/jenkinsci/gradle/plugins/jpi/JpiPublishingAndConsumptionTest.groovy#L205-L210

Reply:
> I think this may be a minor oversight of declaring the dependency twice?

Indeed the dependency `implementation 'org:producer:1.0'` was declare twice in the test as there are two `dependencies { }` blocks added in the test (if you look at `consumerBuild << ...` block a bit further up). But in general you would have to declare the dependency without capability as well. Otherwise the main variant is not selected at all. Which is not always a problem in practice as all variants share the same `jar` in this setup. But if there are transitive dependencies of the main variant that you need at compile time, you need the dependency declaration.  (Although, the JPI manifest is not influenced by that.)

I extended the test a bit to be more clear about what ends up where.